libr/core/cmd_api.c

```diff
@@ -30,12 +56,23 @@ R_API RCmd *r_cmd_new () {
 		cmd->cmds[i] = NULL;
 	}
 	cmd->nullcallback = cmd->data = NULL;
+	cmd->root_cmd_desc = create_cmd_desc (NULL, R_CMD_DESC_TYPE_INNER, "");
+	cmd->ht_cmds = ht_pp_new0 ();
 	r_core_plugin_init (cmd);
 	r_cmd_macro_init (&cmd->macro);
 	r_cmd_alias_init (cmd);
 	return cmd;
 }
 
+static void free_cmd_desc_tree(RCmdDesc *cd) {
```

Comment:
Wouldn't recursive free be useful for plugins which register  multiple commands grouped under single inner node?

Reply:
What do you mean?

Reply:
I think we agreed that in future plugins should prefer the new cmd for registering their commands. That means plugins should be also able to remove and free their commands. I initially thought that calling free_cmd_desc_tree would convenient way of cleaning up. That is during initialization plugin registers "plug" "pluga" and "plugc" and then during cleanup calls recursive free on "plug".

After thinking about this more that's not exactly right API. Better API would be having `cmd_remove(RCmd*, RCmdDesc*)` or `cmd_remove(RCmd*, const char*)` method which does all the cleanup. It should not only free RCmdDesc but also data from other structures like `ht_cmds` and pointer in the parent command.

Reply:
yes, I think everything should pass through RCmd.